jax/experimental/maps.py

```diff
@@ -1126,122 +1127,85 @@ def restore_units(is_unit, vals):
-    args_no_units, in_units = filter_units(args)
-    f, out_units = hide_units(f, tuple(in_units))
-    f, out_named_shapes = out_local_named_shapes(f, frozenset(global_axis_sizes))
-    out_axes_thunk = params['out_axes_thunk']
-    @as_hashable_function(closure=out_axes_thunk)
-    def new_out_axes_thunk():
-      out_axes = out_axes_thunk()
-      axes_units, const_units = split_list(out_units(), [len(out_axes)])
-      assert not any(const_units)
-      num_consts = len(const_units)
-      out_axes_no_units = [a for a, u in zip(out_axes, axes_units) if not u]
-      const_axes: Sequence[AxisNamePos]
-      if num_consts == 0:
-        const_axes = ()
-      else:
-        const_axes = [
-            AxisNamePos(zip(sort_named_shape, range(len(sort_named_shape))),
-                        user_repr=f'<internal: {sort_named_shape}>')
-            for named_shape in out_named_shapes()[-num_consts:]
-            # We sort here to make the iteration order deterministic
-            for sort_named_shape in [sorted(named_shape, key=str)]
-        ]
-      if not const_axes_s:  # NOTE: This can be called multiple times
-        const_axes_s.store(const_axes)
-      assert const_axes_s.val == const_axes
-      return (*out_axes_no_units, *const_axes)
-    pe_params = dict(
-        params,
-        in_axes=tuple(a for a, u in zip(in_axes, in_units) if not u),
-        donated_invars=tuple(a for a, u in zip(donated_invars, in_units) if not u),
-        out_axes_thunk=new_out_axes_thunk)
-    outs_no_units = primitive.bind(f, *args_no_units, **pe_params)
-    new_out_axes_thunk()  # Make sure it is called at least once to compute const_axes
-    return restore_units(out_units(), outs_no_units)
-
-  jaxpr, out_pvals, consts, env_tracers = self.partial_eval(
-      f, in_pvals, app, instantiate=False)
-
-  out_axes = params['out_axes_thunk']()
-  const_axes = const_axes_s.val
+  out_axes_thunk = params['out_axes_thunk']
+
+  # Adjust input tracers' pvals for mapped axes, and unpack.
+  in_pvals = [t.pval if t.pval.is_known() else
+              pe.PartialVal.unknown(
+                  _delete_aval_axes(t.pval.get_aval(), axes, global_axis_sizes))
+              for t, axes in zip(tracers, in_axes)]
+  in_knowns, in_avals, in_consts = pe.partition_pvals(in_pvals)
+
+  # Wrap f to perform partial evaluation, and plumb out aux data.
+  f = pe.trace_to_subjaxpr_nounits(f, self.main, False)
+  f, aux = pe.partial_eval_wrapper_nounits(f, tuple(in_knowns), tuple(in_avals))
+  # Also grab the local named shapes of the output (known and res).
+  f, out_named_shapes = out_local_named_shapes(f, frozenset(global_axis_sizes))
+
+  # Adjust params for knowns (donated_invars, in_axes, out_axes_thunk).
+  @as_hashable_function(closure=out_axes_thunk)
+  def new_out_axes_thunk():
+    out_knowns, _, _, _ = aux()
+    _, out_axes = partition_list(out_knowns, out_axes_thunk())
+    return (*out_axes, *res_axes())
+  def res_axes():
+    _, _, jaxpr, _ = aux()
```

Comment:
Good idea. (There's only `jax_unknown` here, since the known part is evaluated as it runs and there's no jaxpr formed for it. Still it's good to be less ambiguous.)